src/Nonlinear/types.jl

```diff
@@ -0,0 +1,91 @@
+@enum(
+    NodeType,
+    # Index into multivariate operators
+    NODE_CALL_MULTIVARIATE,
+    # Index into univariate operators
+    NODE_CALL_UNIVARIATE,
+    # Index into logic operators
+    NODE_LOGIC,
+    # Index into comparison operators
+    NODE_COMPARISON,
+    # Index is the value of `MOI.VariableIndex`. This is from the original
+    # model, and is not consecutive.
+    NODE_MOI_VARIABLE,
+    # Index of the internal, consecutive, and ordered `MOI.VariableIndex`.
+    NODE_VARIABLE,
+    # Index is into the list of constants
+    NODE_VALUE,
+    # Index is into the list of parameters
+    NODE_PARAMETER,
+    # Index is into the list of subexpressions
+    NODE_SUBEXPRESSION,
+)
+
+struct Node
+    type::NodeType
+    index::Int
+    parent::Int
+end
+
+struct NonlinearExpression
```

Comment:
```suggestion
struct NonlinearExpression <: MOI.AbstractScalarFunction
```

Reply:
I considered this. But if it contains `PARAMETER` or `SUBEXPRESSION` nodes, or a user-defined function then we're in trouble.

I'm going to avoid this for now and see what happens later.